include/pybind11/cast.h

```diff
@@ -78,20 +79,67 @@ PYBIND11_NOINLINE inline internals &get_internals() {
         );
         internals_ptr->static_property_type = make_static_property_type();
         internals_ptr->default_metaclass = make_default_metaclass();
+        internals_ptr->instance_base = make_object_base_type(internals_ptr->default_metaclass);
     }
     return *internals_ptr;
 }
 
-PYBIND11_NOINLINE inline detail::type_info* get_type_info(PyTypeObject *type) {
+/** Takes a Python type and returns the pybind11 type_info for it (if it is a pybind11-registered
+ * type).  If it is not (i.e. it is a python type) this walks the type's inheritance tree to any
+ * inherited pybind11-registered types, which are returned in a vector. (There can be multiple if
+ * an involved python type inherits from multiple pybind11-registered types). */
+PYBIND11_NOINLINE inline std::vector<detail::type_info *> get_all_type_info(PyTypeObject *type) {
+    std::vector<PyTypeObject *> check(1, type);
+    std::vector<detail::type_info *> all_type_info;
     auto const &type_dict = get_internals().registered_types_py;
-    do {
+    for (size_t i = 0; i < check.size(); i++) {
+        type = check[i];
+        if (!PyType_Check((PyObject *) type)) continue;
```

Comment:
I believe that yes, it only applies to Python 2, old-style classes.  (This same logic was essentially there before, though rather indirectly).